plugins/pur/core/src/main/java/org/pentaho/di/repository/pur/PurRepository.java

```diff
@@ -2449,29 +2710,44 @@ protected void saveRepositoryElement( RepositoryElementInterface element, String
     boolean isUpdate = ( element.getObjectId() != null );
     RepositoryFile file;
     if ( isUpdate ) {
-      file = pur.getFileById( element.getObjectId().getId() );
-      // update title & description
-      file =
+      readWriteLock.writeLock().lock();
+      try {
+        file = pur.getFileById( element.getObjectId().getId() );
+
+        // update title & description
+        file =
           new RepositoryFile.Builder( file ).title( RepositoryFile.DEFAULT_LOCALE, element.getName() ).description(
-              RepositoryFile.DEFAULT_LOCALE, Const.NVL( element.getDescription(), "" ) ).build();
+            RepositoryFile.DEFAULT_LOCALE, Const.NVL( element.getDescription(), "" ) ).build();
 
-      // first rename, it is safe as only a name is changed, but not a path
-      renameIfNecessary( element, file );
-      file =
+        // first rename, it is safe as only a name is changed, but not a path
+        renameIfNecessary( element, file );
+
+        file =
           pur.updateFile( file,
-              new NodeRepositoryFileData( transformer.elementToDataNode( element ) ), versionComment );
+            new NodeRepositoryFileData( transformer.elementToDataNode( element ) ), versionComment );
+      } finally {
+        readWriteLock.writeLock().unlock();
+      }
+
     } else {
-      file =
+      readWriteLock.writeLock().lock();
```

Comment:
Here having two locked blocks because they belong to different branches. Could be the same lock however for brevity of code.